beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
                 }
             }
         }
+
+        //
+        // Add missed non-finalized blocks for each monitored validators //
+        //
+        // Count the amount of times a monitored validator missed a non-finalized block
+
+        // We need to extract each validator from the missed_blocks variable and initialise it
+        // to 0. We use a Prometheus gauge and so, this step is mandatory in order to only incr the
+        // counter for non-finalized epochs.
+        let mut validators_missed_blocks: HashMap<u64, u64> = self
+            .missed_blocks
+            .iter()
+            .map(|(_, validator_index, _)| (*validator_index, 0))
+            .collect();
+
+        // Count the amount of times a monitored validator missed block for the non-finalized epochs
+        let finalized_epoch = state.finalized_checkpoint().epoch;
+        self.missed_blocks
+            .iter()
+            .filter(|(epoch, _, _)| *epoch > finalized_epoch)
+            .for_each(|(_, validator_index, _)| {
+                *validators_missed_blocks
+                    .entry(*validator_index)
+                    .or_default() += 1;
+            });
+
+        // Set the Prometheus metrics for each validator (including missed_blocks_count=0)
+        validators_missed_blocks
+            .iter()
+            .for_each(|(validator_index, missed_blocks_count)| {
+                self.aggregatable_metric(validator_index.to_string().as_str(), |label| {
+                    metrics::set_int_gauge(
+                        &metrics::VALIDATOR_MONITOR_MISSED_NON_FINALIZED_BLOCKS_TOTAL,
+                        &[label],
+                        u64_to_i64(*missed_blocks_count),
+                    );
+                });
+            });
+
+        // Increment the Prometheus metrics counter for each missed block
+        if let Some(i) = self.last_missed_block_validator {
```

Comment:
I think the `last_missed_block_validator` has a two issues:

1. It's possible that we can process a state where more than one of our validators have missed a slot. For example, there might be two skipped slots prior to some block and we might be monitoring both of the validators who missed those slots. In this case, `last_missed_block_validator` would contain the second missed proposal but not the first.
1. It's possible that we can process two different states which include the same skipped slot (see example below). In this case `last_missed_block_validator` would register the skipped slot twice.

I'll make another comment further down describing what I think is an easy fix.

### Example of duplicate skipped slots between two blocks

Consider the following block graph:

```
|S0|   |S1|   |S2|    |S3|

  |<---[❌]<--[❌]<--[✅]
  |
[✅]
  |
  |<---[❌]<--[✅]

|S0|   |S1|   |S2|    |S3|
```

- Blocks are represented by ✅ and missed blocks are ❌.
- There are two competing chains. They both agree on slot 0 and 1, but they disagree on slot 2.
- The top chain has missed blocks at slots 1 and 2.
- The bottom chain has missed blocks at only slot 1.
- When importing each of the competing chains we'll detect the missed slot 1.

Reply:
That makes great sense! Thank for the clear explanation!